api/src/main/java/org/openmrs/module/reporting/common/SqlIterator.java

```diff
@@ -0,0 +1,98 @@
+package org.openmrs.module.reporting.common;
+
+import org.openmrs.module.reporting.dataset.DataSetColumn;
+import org.openmrs.module.reporting.dataset.DataSetRow;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SqlIterator implements Iterator<DataSetRow> {
+    private ResultSet resultSet;
+    private List<DataSetColumn> columns;
+    private Statement statement;
+    private Connection connection;
+    private static final Logger log = LoggerFactory.getLogger(SqlIterator.class);
+
+
+    public SqlIterator(ResultSetMetaData metadata, ResultSet resultSet, Connection connection, Statement statement) throws SQLException {
+        this.resultSet = resultSet;
+        this.connection = connection;
+        this.statement = statement;
+        this.createDataSetColumns(metadata);
+    }
+
+    @Override
+    public boolean hasNext() {
+        try {
+            if (resultSet.isLast()) {
```

Comment:
Can you also take a look at this? https://rules.sonarsource.com/java/tag/performance/RSPEC-2232

Reply:
I didn't know that, thanks for the input.
But in this case I don't think that I have any other option, since if I use `next()` I will move the next iteration and that is not what is expected from the `hasNext()`.
Please let me know if you have a better approach for this.

Reply:
How about just throwing UnsupportedOperationException?

Reply:
For the method `hasNext()`? 
I'm pretty sure that the `hasNext()` is used by the `foreach`. 
So that should not be a friendly solution.

Reply:
Do we have to use `foreach`?

Reply:
We don't, but we are. 
And I think that will be very limiting to do that.

See [here](https://github.com/openmrs/openmrs-module-reporting/blob/master/omod/src/main/java/org/openmrs/module/reporting/web/taglib/FormatTag.java#L327) for exemple.

Reply:
@icrc-loliveira for the example you have given, we can add another `if (dataSet instanceof IterableSqlDataSet)` just like we have the `if (dataSet instanceof MapDataSet)`

Reply:
Hello @dkayiwa,
I have removed the use of the `rs.isLast()`.